Include onchain transactions in events

tnull opened the pull request:
Closes #446.
Based on #432.

In #432 we exposed transactions in store. Here we take a stab at exposing them via events. However, to avoid emitting duplicate events for channel-related transactions, this is in draft until we have https://github.com/lightningdevkit/rust-lightning/issues/3566

tnull committed:
Give Wallet access to EventQueue

Let wallet sync enqueue on-chain payment events directly.

Co-Authored-By: HAL 9000

ovitrif commented:
> While the latest BDK 2.2 now shipped event (which would make emitting events easier on our end), we still need [lightningdevkit/rust-lightning#3566](https://github.com/lightningdevkit/rust-lightning/issues/3566) to properly classify the transactions in our store. So this remains blocked until the latter ships (likely with LDK 0.3).

**Is there a way we could get this into ldk-node sooner?**
To be honest we (Bitkit team), kind of relied on the expectation that the BDK 2.2 PR [#310 Return wallet events when applying updates](https://github.com/bitcoindevkit/bdk_wallet/pull/310) will be integrated in ldk-node soon enough (ie. this month or November); and we made our roadmapping around this expectation. Now I'm getting feedback that there is no way we can delay our release any further than when it's planned, and the due date is very soon; which leaves us with very few options:
1. Get help from you to get this into ldk-node very soon (November)
2. Collaborate with you to have someone from our team help with implementation efforts (for example by fixing [lightningdevkit/rust-lightning#3566](https://github.com/lightningdevkit/rust-lightning/issues/3566)), but then we will need a commitment that if we fix that, this PR takes priority and it is implemented shortly thereafter)
3. Do our own implementation for this

The team seems to lean towards favoring option **3**, which is not ideal IMHO, but the main argument is that we don't really have the time to go any other route.

The issue that I see from my PoV is that option **2** might even seem faster.

Could **2** be an option?

tnull committed:
Emit settled on-chain payment events

Notify users when otherwise unclassified on-chain payments reach the
anti-reorg confirmation depth.

Use the wallet's effective stored payment so classified funding,
splice, close, and sweep transactions stay quiet.

Co-Authored-By: HAL 9000

tnull committed:
Return effective values from store upserts

Let callers obtain the object produced by merge semantics without a
second lookup. Preserve the existing boolean API so unrelated call
sites and their allocation behavior remain unchanged.

Co-Authored-By: HAL 9000

tnull committed:
Keep store unit test out of integration tests

The in-memory store source is shared by the library test utilities and
integration-test helpers. Keep its unit test in the library-only module
so each integration binary does not register and rerun it.

Co-Authored-By: HAL 9000

tnull commented:
Should be good for review now.

jkczyz commented on src/data_store.rs:
```diff
@@ -403,6 +459,28 @@ mod tests {
+			logger,
+		);
+
+		let supplied = MergingTestObject { id, data: [25u8; 3], preserved_data: [26u8; 3] };
+		let expected = MergingTestObject { data: supplied.data, ..existing };
+		assert_eq!(Ok((true, expected)), data_store.insert_or_update_and_get(supplied).await);
+
+		let unchanged = MergingTestObject { preserved_data: [27u8; 3], ..expected };
+		assert_eq!(Ok((false, expected)), data_store.insert_or_update_and_get(unchanged).await);
```
Add an assertion for when an insert occurs.

tnull replied:
Done

jkczyz commented on src/wallet/mod.rs:
```diff
@@ -278,16 +278,22 @@ impl Wallet {
-					if payment_status == PaymentStatus::Pending {
-						let pending_payment =
-							self.create_pending_payment_from_tx(payment, Vec::new());
+						if updated && payment_status == PaymentStatus::Succeeded {
```
If we crash after the `payment_store` write but before the event queue is persisted, `updated` will be false on the next sync upon restart and the event will be lost.

Also from Claude:

This `updated` gate is the only thing preventing re-emission after a restart (the reloaded store merges the re-derived payment as a no-op), and no test covers it: no restart test syncs and asserts `next_event() == None`. I wrote that test — it passes, so this is coverage-only.

tnull replied:
Yeah, we applied a somewhat related fix in #948. Maybe it makes sense to first land that then, so we can follow/reuse the exact same pattern here.

jkczyz commented on src/wallet/mod.rs:
```diff
@@ -278,16 +278,22 @@ impl Wallet {
+							let pending_payment =
+								self.create_pending_payment_from_tx(stored_payment, Vec::new());
+							self.pending_payment_store.insert_or_update(pending_payment).await?;
+						}
```
Directly from Claude:

Direct graduation leaves a pre-existing pending entry stale. Sequence: a tx confirms below `ANTI_REORG_DELAY`, so a pending entry is written recording its confirmation block; a shallow reorg then moves the tx into a different block; by the next batch sync the tx is ≥ `ANTI_REORG_DELAY` deep. That sync's `TxConfirmed` graduates the payment and emits an event with the new (canonical) block, but the pending entry still holds the old, now-orphaned block — and `ChainTipChanged` then merges that stale entry back in (`updated == true`), emitting a duplicate event carrying the orphaned block hash and regressing the stored confirmation to the orphaned block. Demonstrated by a failing integration test using a 1-block reorg; no restart needed. Remove/refresh the pending entry when graduating directly — with that fixed, emission is exactly-once for reorgs shallower than `ANTI_REORG_DELAY`, which is worth documenting on the event variants.

jkczyz commented on src/event.rs:
```diff
@@ -271,6 +271,46 @@ pub enum Event {
+		payment_id: PaymentId,
+		/// The transaction identifier.
+		txid: Txid,
+		/// The value, in thousandths of a satoshi, that was sent.
+		amount_msat: u64,
```
Should we include a field for the fees paid?

tnull replied:
Done.

jkczyz commented on src/event.rs:
```diff
@@ -271,6 +271,46 @@ pub enum Event {
+		amount_msat: u64,
+		/// The hash of the block in which the transaction was confirmed.
+		block_hash: BlockHash,
+		/// The height at which the block was confirmed.
```
The height of the block in which the transaction was confirmed.

tnull replied:
Done.

jkczyz commented on src/event.rs:
```diff
@@ -271,6 +271,46 @@ pub enum Event {
+		amount_msat: u64,
+		/// The hash of the block in which the transaction was confirmed.
+		block_hash: BlockHash,
+		/// The height at which the block was confirmed.
```
The height of the block in which the transaction was confirmed.

tnull replied:
Done.

jkczyz commented on src/event.rs:
```diff
@@ -271,6 +271,46 @@ pub enum Event {
 	},
+	/// A sent on-chain payment was successful.
+	///
+	/// This is only emitted for wallet transactions which were not classified as channel
+	/// funding, splices, closes, sweeps, or other LDK-driven chain activity.
```
Claude:

Nit: two narrow cases can still emit these events for LDK-driven transactions and may deserve a doc sentence: (1) classification only happens when our own broadcaster handles a tx, so a closing tx broadcast solely by the counterparty that confirms while this node is offline is discovered by wallet sync unclassified and emits `OnchainPaymentReceived` for the close output; (2) records persisted by pre-classification versions deserialize with `tx_type: None`, so an old LDK-driven tx that only settles after the upgrade emits too. Normal counterparty-initiated coop closes are fine — both peers broadcast the closing tx, so it's classified.

jkczyz commented on tests/common/mod.rs:
```diff
@@ -1546,12 +1621,18 @@ pub(crate) async fn do_channel_full_cycle<E: ElectrumApi>(
 	generate_blocks_and_wait(&bitcoind, electrsd, 6).await;
 	node_a.sync_wallets().unwrap();
 	node_b.sync_wallets().unwrap();
+	if !allow_0conf {
+		assert_eq!(
+			node_a.expect_onchain_payment_event(OnchainPaymentEvent::Received).await,
+			splice_out_txo.txid,
+		);
+	}
```
Hmmm... Clause says we aren't using `InteractiveFunding` for 0-conf splices.

The gate is correct but subtle — worth a comment: with 0conf, LDK re-broadcasts the splice tx as `TransactionType::Funding` (classified → no event); without, the acceptor only broadcasts `InteractiveFunding`, which classification skips → event fires. The cross-variant inconsistency in the acceptor's record may be worth flagging upstream.

tnull replied:
Done.

jkczyz commented on tests/integration_tests_rust.rs:
The CI failure is here according to Claude:

This ordering is what failed CI (`build (self-hosted, beta)`) once the new assertion below started depending on it: `send()` only queues the broadcast (drained by a background task), and `wait_for_tx` after mining also passes for a mempool-only tx — so the tx can miss the first of the 6 blocks, sit below `ANTI_REORG_DELAY` at the sync, and never produce the event (`expect_onchain_payment_event` can't recover; single 60s `next_event_async`, tests/common/mod.rs:426-434). Swap these two lines, as `onchain_send_receive` does.

jkczyz commented on tests/integration_tests_rust.rs:
```diff
@@ -784,6 +829,8 @@ async fn reorged_onchain_payment_returns_to_unconfirmed() {
 	generate_blocks_and_wait(&bitcoind.client, &electrsd.client, 1).await;
 	node_a.sync_wallets().unwrap();
 	node_b.sync_wallets().unwrap();
+	assert_eq!(node_a.next_event(), None);
+	assert_eq!(node_b.next_event(), None);
```
Claude:

After the reorg sync below, also assert no event, then re-mine to `ANTI_REORG_DELAY` and assert exactly one event fires. Extended that way, it currently fails due to the stale-pending-entry duplicate (src/wallet/mod.rs:289).

tnull replied:
Done.

jkczyz commented on tests/integration_tests_rust.rs:
Claude:

Nothing tests that a restarted node doesn't re-emit. I wrote the test — it passes, so coverage-only.

tnull replied:
Done.

jkczyz commented on bindings/kotlin/ldk-node-jvm/lib/src/test/kotlin/org/lightningdevkit/ldknode/LibraryTest.kt:
```diff
@@ -226,6 +226,28 @@ class LibraryTest {
+        val onchainPaymentReceivedEvent1 = node1.waitNextEvent()
+        println("Got event: $onchainPaymentReceivedEvent1")
+        when (onchainPaymentReceivedEvent1) {
+            is Event.OnchainPaymentReceived -> {
+                assertEquals(txid1, onchainPaymentReceivedEvent1.txid)
+                assertEquals(100000000uL, onchainPaymentReceivedEvent1.amountMsat)
+            }
```
How about testing `OnchainPaymentSuccessful`?

tnull replied:
Done.

jkczyz commented:
Tests mentioned in claude feedback can be found in https://github.com/jkczyz/ldk-node/commit/a7c0269c7e564e1f6bf3a1c2d9dc27e95a96d810

tnull committed:
f Handle premine events in reserve test

Upstream added a multi-node reserve test after the event commit's
original base. Consume its premine events so they do not mask the
channel lifecycle assertions after rebasing.

Co-Authored-By: HAL 9000

tnull committed:
f Cover data store insertion result

Exercise the new helper's insertion path as well as its update and
no-op paths so its returned effective object stays covered.

Co-Authored-By: HAL 9000

tnull committed:
f Expose on-chain payment fees

Include the known transaction fee in successful on-chain payment
events so consumers do not need to look up the payment record. Clarify
the confirmation height and legacy classification documentation too.

Co-Authored-By: HAL 9000

tnull committed:
f Stabilize on-chain event tests

Wait for broadcasts before mining so confirmation-depth assertions
cannot race the broadcaster. Document why zero-conf splice behavior
changes whether the receiving wallet emits an event.

Co-Authored-By: HAL 9000

tnull committed:
f Complete on-chain event coverage

Cover successful on-chain events in the Kotlin full-cycle test and
verify that a handled settled event is not re-emitted after a clean
restart.

Co-Authored-By: HAL 9000

tnull committed:
f Assert reorg emits no payment event

Verify that moving a shallow payment back to unconfirmed does not emit
a settled event. Reconfirmation coverage remains with the deferred stale
pending-entry fix.

Co-Authored-By: HAL 9000

tnull commented:
Hmm, given the reclassification inconsistencies we're still facing the issue that counterparty spends might wrongfully trigger an onchain payment event.

I now explored an alternative approach over at https://github.com/tnull/ldk-node/tree/2025-01-include-onchain-transactions-in-events-alt where we (ab)use the `KeychainKind`, i.e., discern LDK spends and onchain payments via `Internal`/`External` descriptors. That seems like a hacky workaround and we'd also still get misfiring events for addresses already handed out before the upgrade. So not quite sure if it's worth it yet. Will hold a bit longer until we're clear on if we ship further improvements for the upcoming release.

tnull commented:
Moving this off the v0.8 milestone. I increasingly think we need to reconsider how we do onchain transaction classification, and until we do it's probably not feasible to keep piling on masses of hacky code to keep working around the shortcomings of the central architectural choices.
